server/src/main/java/io/crate/planner/optimizer/collect/rule/MoveArrayLengthOnReferenceCastToLiteralCastInsideOperators.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.planner.optimizer.collect.rule;
+
+import io.crate.expression.scalar.ArrayUpperFunction;
+import io.crate.expression.symbol.Function;
+import io.crate.expression.symbol.Symbol;
+import io.crate.expression.symbol.SymbolType;
+import io.crate.metadata.Functions;
+import io.crate.planner.optimizer.collect.Rule;
+import io.crate.planner.optimizer.matcher.Capture;
+import io.crate.planner.optimizer.matcher.Captures;
+import io.crate.planner.optimizer.matcher.Pattern;
+import io.crate.types.DataType;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.function.BiFunction;
+
+import static io.crate.expression.operator.Operators.COMPARISON_OPERATORS;
+import static io.crate.expression.scalar.cast.CastFunctionResolver.CAST_FUNCTION_NAMES;
+import static io.crate.planner.optimizer.matcher.Pattern.typeOf;
+
+
+public class MoveArrayLengthOnReferenceCastToLiteralCastInsideOperators implements Rule<Function> {
+
+    private final Capture<Function> castCapture;
+    private final Pattern<Function> pattern;
+    private final BiFunction<String, List<Symbol>, Symbol> functionResolver;
+    private volatile boolean enabled = true;
+
+    public MoveArrayLengthOnReferenceCastToLiteralCastInsideOperators(
+        BiFunction<String, List<Symbol>, Symbol> functionResolver
+    ) {
+        this.functionResolver = functionResolver;
+        this.castCapture = new Capture<>();
+        this.pattern = typeOf(Function.class)
+            .with(f -> COMPARISON_OPERATORS.contains(f.info().ident().name()))
+            .with(f -> f.arguments().get(1).symbolType() == SymbolType.LITERAL)
+            .with(f -> Optional.of(f.arguments().get(0)), typeOf(Function.class).capturedAs(castCapture)
```

Comment:
Should we try to create some "matcher" functions, similar to how `isFunction(name, argMatcher...)` works that we use in the tests?

Reply:
Good idea, will try.

Reply:
I'd suggest to do this as a follow up so we first could get this functionality merged in and finish up the complete function registry overhaul (move missing functions, afaik only operators left).